### PR TITLE
fix: SOF-1135 fix half removed device

### PR DESCRIPTION
**Pull request by ianshade**
Fixes a bug that caused a device to be half-removed if its creation timed out, leaving it in an unwanted state that required user intervention. Refactors the device update logic, uses a queue to manage the actions. Some actions are broken down into multiple chained jobs to facilitate cleanup if one of the jobs fails or if the queue is cleared entirely (when settings of a device change before it's fully created and initialized, or a device pending creation is no longer needed)
Depends on https://github.com/tv2/tv-automation-state-timeline-resolver/pull/44

### CODE REVIEW

File: packages/playout-gateway/src/tsrHandler.ts [hunk 1]
```diff
@@ -597,407 +587,241 @@ export class TSRHandler {
...
+				() => {
+					this.updateDeviceStatus(deviceId, JobFailure.ADD_TIMEOUT)
+					this._triggerUpdateDevices()
+				}
+			)
+	}
+
+	private enqueueSetDebugLogging(deviceId: string, debugLogging: boolean) {
+		const jobQueue = this._jobQueueManager.get(deviceId)
+		const jobId = 'updateExpectedPlayoutItems'
```
Comment (KvelaGorrrrnio) [inline packages/playout-gateway/src/tsrHandler.ts:696]: Is this the correct `jobId`?

File: packages/playout-gateway/src/tsrHandler.ts [hunk 2]
```diff
@@ -1013,75 +837,18 @@ export class TSRHandler {
...
-							}
-						}
-					)
-				)
+			const { groupedExpectedItems, rundowns } = this.getExpectedPlayoutItems()
+			this.tsr.getDevices().forEach((device) => {
+				this.enqueueExpectedPlayoutItemsUpdate(groupedExpectedItems, rundowns, device.deviceId)
 			})
-		)
+		}, 200)
```
Comment (KvelaGorrrrnio) [inline packages/playout-gateway/src/tsrHandler.ts:851]: Perhaps this should be extracted into a constant.

File: packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+			})
+
+			this.tsrHandler.onDeviceChildClosed(this.deviceId)
+		}
+		// Note for the future:
+		// It is important that the callbacks returns void,
+		// otherwise there might be problems with threadedclass!
+
+		await deviceContainer.device.on('connectionChanged', onDeviceStatusChanged as () => void)
+		// await device.device.on('slowCommand', onSlowCommand)
```
Comment (KvelaGorrrrnio) [inline packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts:160]: Should this be removed?

File: packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+		await deviceContainer.device.on('debug', (...args: any[]) => {
+			if (!deviceContainer.debugLogging && !this.tsrHandler.coreHandler.logDebug) {
+				return
+			}
+			if (args.length === 0) {
+				this.logger.debug('>empty message<')
+				return
+			}
+			const data = args.map((arg) => (typeof arg === 'object' ? JSON.stringify(arg) : arg))
```
Comment (KvelaGorrrrnio) [inline packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts:186]: Parantheses around `typeof arg === 'object' ? JSON.stringify(arg) : arg` are unnecessary.

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+		job: Job<T, void, V>,
+		timeout?: number,
+		importance?: JobImportance,
+		id?: string
+	): IChainableJobQueue<T>
+	removeJob(jobId: string): void
+}
+
+export class JobQueue extends EventEmitter implements IJobQueue {
+	private queue: JobDescription<any>[] = []
```
Comment (KvelaGorrrrnio) [inline packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:49]: We should avoid using `any` if possible.

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+	}
+
+	clear(): void {
+		if (this.currentJob) {
+			this.abortJob(this.currentJob)
+		} else {
+			if (this.currentJobChain.length) {
+				this.currentJob = this.currentJobChain[this.currentJobChain.length - 1]
+
+				Promise.allSettled(this.currentJobChain.map(async (job) => job.job.cleanup()))
```
Comment (KvelaGorrrrnio) [inline packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:69]: `async` is redundant here.

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+	}
+
+	get length(): number {
+		return this.queue.length
+	}
+
+	clear(): void {
+		if (this.currentJob) {
+			this.abortJob(this.currentJob)
+		} else {
```
Comment (KvelaGorrrrnio) [inline packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:65]: By adding `return` inside the if-statement, the else-branch can be dissolved to decrease the number of nested levels.

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+			return this.currentJob.id
+		}
+		return undefined
+	}
+
+	removeJob(jobId: string): void {
+		this.queue = this.queue.filter((job) => job.id !== jobId && !job.chained)
+	}
+
+	endsWith(jobId: string): boolean {
```
Comment (KvelaGorrrrnio) [inline packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:122]: Suggestion: Rename to `isLastJob` to match syntax for `getLastJobIdByImportance`.

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+		const lastJob = this.queue[this.queue.length - 1] ?? this.currentJob
+		return lastJob?.id === jobId
+	}
+
+	private run(): void {
```
Comment (KvelaGorrrrnio) [inline packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:127]: Should `run` not make use of the importance, when choosing a job from the queue?

File: packages/playout-gateway/src/coreHandler.ts [hunk 10]
```diff
@@ -653,16 +646,19 @@ export class CoreTSRDeviceHandler {
 			)
 	}
 
-	async dispose(): Promise<void> {
+	async dispose(expected: boolean): Promise<void> {
```
Comment (andr9528) [inline packages/playout-gateway/src/coreHandler.ts:649]: The name `expected` doesn't immediatly give me an idea of what the argument is used for.
It definetly doesn't make it obvious that it decides between `StatusCode.UNKNOWN` and `StatusCode.BAD`.

Renaming it to e.g. `expectedStatusCode`, would give an indication that the boolean effects a statuscode. Of cource simply changing the type from `boolean` to `StatusCode.UNKNOWN | StatusCode.BAD` would make it very obvivous what the expected is, and move the chose out to where the dispose is getting called.

Comment (LindvedKrvang) [reply packages/playout-gateway/src/coreHandler.ts:649]: If you rename it to `expectedStatusCode` then the name indicates that it is a StatusCode and not a boolean.

File: packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+
+		const { deviceContainer, coreTsrHandler } = previousResult
+		// Set up device status
+		const deviceType = deviceContainer.deviceType
+
+		const onDeviceStatusChanged = (connectedOrStatus: Partial<DeviceStatus>) => {
+			let deviceStatus: Partial<StatusObject>
+			if (_.isBoolean(connectedOrStatus)) {
+				// for backwards compability, to be removed later
+				if (connectedOrStatus) {
```
Comment (LindvedKrvang) [inline packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts:47]: Could be a ternary

File: packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+			} else {
+				deviceStatus = connectedOrStatus
+			}
+			coreTsrHandler.statusChanged(deviceStatus)
+
+			// When the status has changed, the deviceName might have changed:
+			deviceContainer.reloadProps().catch((err) => {
+				this.logger.error(`Error in reloadProps: ${err}`)
+			})
+			// hack to make sure atem has media after restart
```
Comment (LindvedKrvang) [inline packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts:66]: Since it was necessary to leave a comment explaining a portion of the method it indicates that said portion should instead be extracted into a method. Then you don't need the comment because the name of the extracted method will serve as the comment.

File: packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+			onDeviceStatusChanged({
+				statusCode: StatusCode.BAD,
+				messages: ['Child process closed'],
+			})
+
+			this.tsrHandler.onDeviceChildClosed(this.deviceId)
+		}
+		// Note for the future:
+		// It is important that the callbacks returns void,
+		// otherwise there might be problems with threadedclass!
```
Comment (LindvedKrvang) [inline packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts:157]: This makes me a bit sad to see :(

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts [hunk 14]
```diff
@@ -0,0 +1,231 @@
+import { EventEmitter } from 'stream'
+import { AbortError } from 'timeline-state-resolver'
+import { Logger } from 'winston'
+import { Job } from './job'
+
+interface JobDescription<X> {
```
Comment (LindvedKrvang) [inline packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:6]: Any reason for not using `T` instead of `X`?

**Review by LindvedKrvang — APPROVED** [review]

File: packages/playout-gateway/src/tsrHandler.ts [hunk 15]
```diff
@@ -120,44 +119,30 @@ export interface RoutedMappings {
...
 	private _coreHandler!: CoreHandler
 	private _triggerupdateExpectedPlayoutItemsTimeout: any = null
 	private _coreTsrHandlers: { [deviceId: string]: CoreTSRDeviceHandler } = {}
 	private _observers: Array<any> = []
 	private _cachedStudioId = ''
+	private _jobQueueManager: JobQueueManager
+	private _deviceJobFailureStatuses = new Map<string, JobFailure>()
 
 	private _initialized = false
 	private _multiThreaded: boolean | null = null
```
Comment (andr9528) [inline packages/playout-gateway/src/tsrHandler.ts:148]: What benefit does these fields have being of type `boolean | null` over say `boolean | undefined`?
Is a null check not more processing costly than a undefined check?

Also aren't these boolean values missing the `isX`, or `doesX` naming scheme to better indicate that they are booleans?

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+			this.currentJob.onFulfilled(this, result)
+		}
+	}
+
+	private handleJobRejected = async (reason: any): Promise<PromiseSettledResult<void>[] | void> => {
+		if (!this.currentJob) return
+		this.logger.error(
+			`Error in queue "${this.name}" at Job.run "${this.currentJob.job.constructor.name}": ${reason}`
+		)
+		const ps = []
```
Comment (RasmusAlbrektsen) [inline packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:175]: This might just be me, but abbreviations on arrays make it hard to follow the general idea of having a descriptive name in plural. I know Rasmus' made comments about the abbreviation for promises as well on the related PR.

**Review by RasmusAlbrektsen — APPROVED** [review]

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts [hunk 17]
```diff
@@ -0,0 +1,231 @@
+import { EventEmitter } from 'stream'
+import { AbortError } from 'timeline-state-resolver'
+import { Logger } from 'winston'
+import { Job } from './job'
+
+interface JobDescription<X> {
+	id?: string
+	job: Job<X, any, any>
```
Comment (andr9528) [inline packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:8]: Propably nothing/not possibly, but is there really no other type than `any`, that can be applied?

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+
+	private run(): void {
+		setImmediate(() => {
+			if (this.currentJob) {
+				return
+			}
+			if (this.queue.length === 0 && this.currentJobChain.length === 0) {
+				this.emit('done')
+			}
+			this.currentJob = this.queue.shift() ?? null
```
Comment (andr9528) [inline packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:135]: I am pretty sure the Nullish coalescing operator operator is redundant.

If the result of `this.queue.shift()` is null, then currently null is assinged to `this.currentJob` with or without the Nullish coalescing operator.

If a non-null value is to be assigned in the case of a null result from `this.queue.shift()`, then it is needed, but as is, should not be the case.

I am unsure if the Nullish coalescing operator has any effect incase `this.queue.shift()` returns `undefined`, if possibly.

File: packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts (hunk 4 above)
Comment (ianshade) [reply packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts:186]: But they are required by the linter 
image

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+	}
+
+	private handleJobFulfilled = (result: any): void => {
+		if (!this.currentJob) return
+		if (this.currentJob.abortController.signal.aborted) {
+			throw new AbortError()
+		}
+		this.currentJob.result = result
+		const nextJob = this.queue[0]
+		if (nextJob?.chained) {
```
Comment (andr9528) [inline packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:162]: Is `.chained` a boolean or being checked for undefined?
I'd guess a boolean, but i shouldn't need to guess.

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+	}
+
+	private handleJobRejected = async (reason: any): Promise<PromiseSettledResult<void>[] | void> => {
+		if (!this.currentJob) return
+		this.logger.error(
+			`Error in queue "${this.name}" at Job.run "${this.currentJob.job.constructor.name}": ${reason}`
+		)
+		const ps = []
+		if (this.currentJobChain.length) {
+			ps.push(...this.currentJobChain.map(async (job) => job.job.cleanup()))
```
Comment (andr9528) [inline packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:177]: Which in the `job.job` is a job, and which is something else? If my understanding of the code is not wrong, then a job do not contain another job.

On top of this `job.job` looks weird.

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts (hunk 6 above)
Comment (ianshade) [reply packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:69]: I tried but [@typescript-eslint/promise-function-async](https://typescript-eslint.io/rules/promise-function-async) is enabled
image

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts (hunk 8 above)
Comment (ianshade) [reply packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:122]: Better idea: I'll remove it because it is not used

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts (hunk 9 above)
Comment (ianshade) [reply packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:127]: That's why it's called _importance_ - it is not the same as priority, but If you have better ideas for naming, I'm all ears

Comment (ianshade) [reply packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:127]: In this particular use case there should not be a scenario when a job with high importance is in the queue after a job with low importance. We just clear the queue before it can occur

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts (hunk 18 above)
Comment (LindvedKrvang) [reply packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:135]: The result of Array.shif() is never null. It returns the removed element or `undefined`

File: packages/playout-gateway/src/tsrHandlerJobs/job.ts [hunk 21]
```diff
@@ -0,0 +1,18 @@
+interface IJob<OwnResultType, PrevResultType> {
```
Comment (andr9528) [inline packages/playout-gateway/src/tsrHandlerJobs/job.ts:1]: Hmmm something about avoid `I...` interface naming, as it is a standard not used in typescript.

I for one do not mind it, but i know Lindved has mentioned it a bunch of times when i was developing the TricasterIpInfo and also to Ida, as she also have a tendency to add it. You could say it is a work damage, from being mainly a C# dev.

File: packages/playout-gateway/src/coreHandler.ts (hunk 10 above)
Comment (andr9528) [reply packages/playout-gateway/src/coreHandler.ts:649]: > If you rename it to `expectedStatusCode` then the name indicates that it is a StatusCode and not a boolean.

Was also thinking about that after i wrote my comment. Naming it `isExpectedStatusCode`, would indicate it being a boolean. The name has however grown considerably in size, so changing the type as suggested, would propably be better.

[the remark this answers is not in the document]
Comment (andr9528) [reply packages/playout-gateway/src/coreHandler.ts:649]: Lindved's Example show how i'd suggest the input of dispose to be, from my other comment.

Creating a new StatusCode, called `StatuseCode.UNINITIALIZED` could also be a solution, if creating a new statuscode is allowed. In my mind, when something gets disposed, it gets uninitialized, thus having a statuscode that indicates this directly, instead of having to decide between two, might be an alternative solution.

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts (hunk 18 above)
Comment (andr9528) [reply packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:135]: Ahh i see, Then the Nullish coalescing turns the undefined into a null.
I just checked how `.shift()` works in a sandbox, don't know why i didn't before writing it.

That said i think i found a bug in 'codesandbox.io'. It seems to have trouble with understanding a Nullish coalescing operator, seeing/expecting it as two ternary operators.

File: packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts (hunk 16 above)
Comment (andr9528) [reply packages/playout-gateway/src/tsrHandlerJobs/jobQueue.ts:175]: Not just you, i think it was Anders who commented on a similar abbreviations, somewhere else.
Some abbreviations can be ok, in my mind, if it is otherwise a very long name, as long as understanding of what it is, is not lost. - this is not such a case.

File: packages/playout-gateway/src/tsrHandler.ts (hunk 15 above)
Comment (ianshade) [reply packages/playout-gateway/src/tsrHandler.ts:148]: > What benefit does these fields have being of type boolean | null over say boolean | undefined?

Semantically `boolean | undefined` would make more sense.

> Is a null check not more processing costly than a undefined check?

If anything, it might be faster https://www.measurethat.net/Benchmarks/Show/6389/0/compare-comparison-with-null-or-undefined

File: packages/playout-gateway/src/coreHandler.ts [hunk 22]
```diff
@@ -537,19 +536,13 @@ export class CoreTSRDeviceHandler {
...
 
-	constructor(
-		parent: CoreHandler,
-		device: Promise<DeviceContainer<DeviceOptionsAny>>,
-		deviceId: string,
-		tsrHandler: TSRHandler
-	) {
-		this._coreParentHandler = parent
-		this._devicePr = device
+	constructor(device: DeviceContainer<DeviceOptionsAny>, deviceId: string, tsrHandler: TSRHandler) {
```
[the remark this answers is not in the document]
Comment (ianshade) [reply packages/playout-gateway/src/coreHandler.ts]: updated

File: packages/playout-gateway/src/tsrHandlerJobs/createDeviceJob.ts [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+	deviceContainer: DeviceContainer<DeviceOptionsAny>
+	coreTsrHandler: CoreTSRDeviceHandler
+}
+
+export class CreateDeviceJob extends Job<CreateDeviceJobsResult, any, Partial<CreateDeviceJobsResult>> {
+	protected artifacts: Partial<CreateDeviceJobsResult> = {}
+
+	constructor(private deviceId: string, private deviceOptions: DeviceOptionsAny, private tsrHandler: TSRHandler) {
+		super()
+	}
```
[the remark this answers is not in the document]
Comment (ianshade) [reply packages/playout-gateway/src/tsrHandlerJobs/createDeviceJob.ts:16]: updated in most places

File: packages/playout-gateway/src/tsrHandlerJobs/createDeviceJob.ts [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+	async run(_previousResult?: undefined, abortSignal?: AbortSignal): Promise<CreateDeviceJobsResult> {
```
[the remark this answers is not in the document]
Comment (ianshade) [reply packages/playout-gateway/src/tsrHandlerJobs/createDeviceJob.ts]: I think it is improved now, but generally It is a bit tricky to have complete type safety on generic containers that can hold a mix of multiple types not known ahead of time

File: packages/playout-gateway/src/tsrHandler.ts (hunk 1 above)
Comment (ianshade) [reply packages/playout-gateway/src/tsrHandler.ts]: It was not :)

File: packages/playout-gateway/src/tsrHandler.ts (hunk 2 above)
Comment (ianshade) [reply packages/playout-gateway/src/tsrHandler.ts]: Extracted

File: packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts (hunk 3 above)
Comment (ianshade) [reply packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts]: Old old code, yes

File: packages/playout-gateway/src/coreHandler.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (ianshade) [reply packages/playout-gateway/src/coreHandler.ts]: I refactored `dispose`, but the flag still is on the job's constructor because it does seem like an overkill to make two job classes, each calling a different method, so perhaps changing the parameter to a StatuCode would make more sense, but that also seems to be violating some rules too.
I kind of like the idea of a new status code, but I feel like the semantics and reporting logic of it require a lot of thought.
For a regular user looking at the Rundown View, it should be clear whether performing any playout action might fail due to any reason - could be that the Playout gateway has/was restarted, could be that some device is misconfigured etc.. From that perspective these are all bad states and for example if the user does a Take, it will probably look bad on the output. On the other hand, the detailed cause of that state, relevant for a superuser or support, is in the message, but I can see that for the purpose of health monitoring, a distinction between an expected initialization (e.g. due to settings change or a restart triggered by the user) could be useful to assess the severity and need for intervention.

File: packages/playout-gateway/src/tsrHandler.ts [hunk 25]
```diff
@@ -597,407 +590,240 @@ export class TSRHandler {
...
+				failedDevices[JobFailure.REMOVE_ERROR].length > 0
+					? `Failed to remove devices: ${stringifyIds(failedDevices[JobFailure.REMOVE_ERROR])}`
+					: null,
+				failedDevices[JobFailure.ADD_TIMEOUT].length > 0
+					? `Time-out when adding devices: ${stringifyIds(failedDevices[JobFailure.ADD_TIMEOUT])}`
+					: null,
+				failedDevices[JobFailure.REMOVE_TIMEOUT].length > 0
+					? `Time-out when removing devices: ${stringifyIds(failedDevices[JobFailure.REMOVE_TIMEOUT])}`
 					: null,
 			].filter(Boolean) as string[],
```
Comment (andr9528) [inline packages/playout-gateway/src/tsrHandler.ts:804]: Do i understand it correctly, that it trims away any that are `null`?

File: packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts (hunk 11 above)
Comment (andr9528) [reply packages/playout-gateway/src/tsrHandlerJobs/initDeviceJob.ts:48]: +1 to Rasmus' Comment.

``` typescript
const deviceStatus = _.isBoolean(connectedOrStatus) 
  ? connectedOrStatus
    ? { statusCode: StatusCode.GOOD, }
    : {	statusCode: StatusCode.BAD, messages: ['Disconnected'], }
  : connectedOrStatus
```

Making it a double ternary, may or may not be wise.

File: packages/playout-gateway/src/tsrHandlerJobs/removeDeviceJob.ts [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+		private expected: boolean,
+		private logger: Logger
+	) {
+		super()
+	}
+
+	async run(): Promise<void> {
+		if (this.coreTsrHandlers[this.deviceId]) {
+			try {
+				if (this.expected) {
```
Comment (andr9528) [inline packages/playout-gateway/src/tsrHandlerJobs/removeDeviceJob.ts:20]: Could be a ternary.

``` typescript
const dispose = this.expected
  ? this.coreTsrHandlers[this.deviceId].disposeExpectedly
  : this.coreTsrHandlers[this.deviceId].disposeUnexpectedly
await dispose()
```
Should shrink the amount of lines used by one or two lines. 
Could move the call to the same line as the ternary, with use of parenthesis, if wanted.

``` typescript
await (this.expected
  ? this.coreTsrHandlers[this.deviceId].disposeExpectedly
  : this.coreTsrHandlers[this.deviceId].disposeUnexpectedly)()
```

Comment (Julusian) [reply packages/playout-gateway/src/tsrHandlerJobs/removeDeviceJob.ts:20]: Does it benefit from being less lines? 
I think that makes it a lot harder to read

Especially the second version, where if skim reading, you can easily miss that disposeExpectedly is a function
